language/diem-framework/modules/Vote.move

```diff
@@ -0,0 +1,377 @@
+                counter: 0,
+            });
+        };
+        if (!exists<Ballots<Proposal>>(ballot_address)) {
+            move_to(ballot_account, Ballots<Proposal> {
+                ballots: Vector::empty(),
+                create_ballot_handle: Event::new_event_handle<CreateBallotEvent<Proposal>>(ballot_account),
+                remove_ballot_handle: Event::new_event_handle<RemoveBallotEvent>(ballot_account),
+                voted_handle: Event::new_event_handle<VotedEvent>(ballot_account),
+                ballot_approved_handle: Event::new_event_handle<BallotApprovedEvent>(ballot_account),
+            });
+        };
+
+        let ballot_data = borrow_global_mut<Ballots<Proposal>>(ballot_address);
+
+        // Remove any expired ballots
+        gc_internal<Proposal>(ballot_data);
+        let ballots = &mut ballot_data.ballots;
+
+        assert(Vector::length(ballots) < MAX_BALLOTS_PER_PROPOSAL_TYPE_PER_ADDRESS, Errors::limit_exceeded(ETOO_MANY_BALLOTS));
+        let ballot_id = new_ballot_id(incr_counter(ballot_account), ballot_address);
+        let ballot = Ballot<Proposal> {
+            proposal,
+            proposal_type,
+            num_votes_required,
+            allowed_voters,
+            votes_received: Vector::empty(),
+            total_weighted_votes_received: 0,
+            ballot_id: *(&ballot_id),
+            expiration_timestamp_secs,
+        };
+        Vector::push_back(ballots, *(&ballot));
+        Event::emit_event<CreateBallotEvent<Proposal>>(
+            &mut borrow_global_mut<Ballots<Proposal>>(ballot_address).create_ballot_handle,
+            CreateBallotEvent {
+                ballot_id: *(&ballot_id),
+                ballot,
+            },
+        );
+        ballot_id
+    }
+
+    // Checks if a voter is present in the vector<WeightedVoter>
+    fun check_voter_present(
+        weighted_voters: &vector<WeightedVoter>,
+        voter: &vector<u8>,
+    ): bool {
+        let i = 0;
+        let len = Vector::length(weighted_voters);
+        while (i < len) {
+            if (&Vector::borrow(weighted_voters, i).voter == voter) {
+                return true
+            };
+            i = i + 1;
+        };
+        false
+    }
+
+    /// Submit a vote from the `voter_account` to `ballot_id`
+    /// This also contains the `proposal_type` and `proposal` so that
+    /// the voter signs over these when sending their vote.
+    /// If this vote causes the ballot to be approved, then the ballot
+    /// is removed from the proposers address.
+    /// This returns a bool indicating whether this vote moved the ballot
+    /// to an approved status. true represents that the ballot got approved
+    /// after this vote and false represents that the ballot has not been
+    /// approved after this vote
+    public fun vote<Proposal: store + drop>(
+        voter_account: &signer,
+        ballot_id: BallotID,
+        proposal_type: vector<u8>,
+        proposal: Proposal,
```

Comment:
I wonder if requiring the `Proposal` value here is too high of a bar. The particular scenario I'm thinking about is if one of the values in the `Proposal` is guarded/hard to construct/meant to be private, then creating one to vote or being forced to "open up the type" so that it can be used in a `Proposal` to be voted on seems like it could be a bit hindrance. Another possibility might be to allow a ballot specify if the whole `Proposal` needs to be provided in a vote or not.

Reply:
That's a good point.
On one hand, since `ballot_id` is unique, it's hard to see a way for a vote to be maliciously reused for non-intended proposals.
On the other hand, a voting agent, which needs to construct a voting transaction, is ultimately required to show the specific proposal to a client, which would most probably require some proposal visualization, explicitly signed by the client custody.

I'm not concerned about opening up the types, because IIUC, it's the responsibility of the client module to design a specific proposal struct for signing. In most cases it should be a struct with a bunch of parameters (e.g., `CreateValidatorProposal {pubkey, human_name}`).

Reply:
In the scenario where the signing client does not interact with a trusted full node, there is a possibility of the signing client being shown a proposal which does not match the proposal onchain. 

```
signing client (eg: iOS App) -> intermediate helper service (untrusted/semi-trusted) -> blockchain
```

In this case, the client would end up signing only the ballot id for a proposal which does not match the proposal onchain.

> one of the values in the Proposal is guarded/hard to construct/meant to be private

I don't think this will be a problem. A module M which is going to use the vote module, will typically have two functions
* `public(script) fun create_proposal_for_x(params1...)` which will call `Vote::create_ballot(proposal: Proposal, ....)`
* `public(script) fun vote_on_proposal_for_x(params2...)` which will call `Vote::vote(proposal: Proposal, ....)`

The Proposal should be constructible from params1 and if that is the case, it should be constructible from params2 as well.

Reply:
Thanks for the response and for the discussion on this -- especially with you're first point, I'm convinced that this is the right way to go (especially since `Proposal` is `store + drop`).